### PR TITLE
Update voltha

### PR DESCRIPTION
**Pull request by uottremb**
Update Voltha script with multiple BBSim capabilities (VOL-2621)

### CODE REVIEW

File: voltha [hunk 1]
```diff
@@ -108,6 +110,18 @@ if [ $HOSTARCH == "x86_64" ]; then
...
 
+# check number (range) of bbsim, max bbsim must not exceed 10 instances!
+# note: instances will be numbered from 0 to 9
+if [ "$1" == "up" ]; then
+	if [ "$NUM_OF_BBSIM" -le 1 ]; then 
+			echo -e "   \033[1m\033[5m\033[31mSetting of BBSIM instances to minimum instances of 1!\033[0m"
+			NUM_OF_BBSIM=1
+	fi
+	if [ "$NUM_OF_BBSIM" -ge $(($MAX_NUM_OF_BBSIM)) ]; then 
+			echo -e "   \033[1m\033[5m\033[31mSetting of BBSIM instances to max supported instances of $MAX_NUM_OF_BBSIM!\033[0m"
```
Comment (Andrea-Campanella) [inline voltha]: what is `\033[1m\033[5m\033[31m`

File: voltha [hunk 2]
```diff
@@ -108,6 +110,18 @@ if [ $HOSTARCH == "x86_64" ]; then
     HOSTARCH="amd64"
 fi
 
+# check number (range) of bbsim, max bbsim must not exceed 10 instances!
+# note: instances will be numbered from 0 to 9
+if [ "$1" == "up" ]; then
+	if [ "$NUM_OF_BBSIM" -le 1 ]; then 
+			echo -e "   \033[1m\033[5m\033[31mSetting of BBSIM instances to minimum instances of 1!\033[0m"
```
Comment (Andrea-Campanella) [inline voltha]: what are `\033[1m\033[5m\033[31m`

Comment (Andrea-Campanella) [reply voltha]: also strange chars present at the end of the string

File: voltha (hunk 1 above)
Comment (Andrea-Campanella) [reply voltha]: also at the end

**Comment by hwindlass** [comment]
Please update README as well and mention that NUM_OF_BBSIM is used more like a sub-option to existing WITH_BBSIM flag.

File: voltha (hunk 2 above)
Comment (uottremb) [reply voltha]: this 'strange' signs are format signs for the bash, if the given bbsim instances out of the defined range, I overwrite the settings and I think, the user should informed about it. The first characters make the following words blink red, the last characters switch back to the normal mode of the bash.

Comment (Andrea-Campanella) [reply voltha]: everyday you learn something new. Thanks!

**Comment by uottremb** [comment]
Update Voltha script with multiple BBSim capabilities

**Comment by uottremb** [comment]
sorry, I try to learn github ;)

File: README.md [hunk 3]
```diff
@@ -90,6 +90,7 @@ Please check the `releases` folder to see the available ones.
 | `VOLTHA_ADAPTER_OPEN_OLT_CHART` | onf/voltha-adapter-openolt   | Helm chart to use to install OpenOlt adapter                                        |
 | `VOLTHA_ADAPTER_OPEN_ONU_CHART` | onf/voltha-adapter-openonu   | Helm chart to use to install OpenOnu adapter                                        |
 | `ONLY_ONE`                      | yes                          | Run a single `rw-core`, no `api-server`, and no `ssh` CLI                           |
+| `NUM_OF_BBSIM`                  | 1                            | number of BBSIM POD to start (minimum = 1, maximum = 10)                            |
```
Comment (hwindlass) [inline README.md]: Minor suggestion: Move this is up under WITH_BBSIM or mention here to use this in conjunction with the other. Because default of WITH_BBSIM is NO and setting this alone is of no good.
Rest looks fine to me.

File: voltha [hunk 4]
```diff
@@ -100,6 +100,8 @@ VOLTHA_ADAPTER_OPEN_ONU_CHART=${VOLTHA_ADAPTER_OPEN_ONU_CHART:-onf/voltha-adapte
 VOLTHA_ADAPTER_OPEN_ONU_CHART_VERSION=${VOLTHA_ADAPTER_OPEN_ONU_CHART_VERSION:-latest}
 ONOS_CHART_VERSION=${ONOS_CHART_VERSION:-latest}
 EXTRA_HELM_INSTALL_ARGS=
+NUM_OF_BBSIM=${NUM_OF_BBSIM:-1}
+MAX_NUM_OF_BBSIM=${MAX_NUM_OF_BBSIM:-10}
```
Comment (dbainbri-ciena) [inline voltha]: Why is max configurable?

File: voltha [hunk 5]
```diff
@@ -108,6 +110,18 @@ if [ $HOSTARCH == "x86_64" ]; then
     HOSTARCH="amd64"
 fi
 
+# check number (range) of bbsim, max bbsim must not exceed 10 instances!
+# note: instances will be numbered from 0 to 9
+if [ "$1" == "up" ]; then
+	if [ "$NUM_OF_BBSIM" -le 1 ]; then
```
Comment (dbainbri-ciena) [inline voltha]: seems like the wrong comparison, as if NUM_OF_BBSIM is 1 it will warn that it it setting to minimum of 1. don't you mean -lt

File: voltha (hunk 2 above)
Comment (dbainbri-ciena) [inline voltha]: voltha up defines constants for color and format these should be used instead of the raw character codes. serach for RED in the script.

File: voltha (hunk 1 above)
Comment (dbainbri-ciena) [reply voltha]: do you me just -gt, as this will trigger if NUM == MAX

File: voltha [hunk 6]
```diff
@@ -605,7 +619,7 @@ if [ "$1" == "down" ]; then
         fi
     else
         EXISTS=$(helm list -q)
-        EXPECT="etcd-operator onos open-olt open-onu sim voltha bbsim radius"
+        EXPECT="etcd-operator onos open-olt open-onu sim voltha bbsim bbsim1 bbsim2 bbsim3 bbsim4 bbsim5 bbsim6 bbsim7 bbsim8 bbsim9 radius"
```
Comment (dbainbri-ciena) [inline voltha]: not confident this will work as what if i set the number to 5?

**Comment by dbainbri-ciena** [comment]
It looks like you may be installing a separate deployment for each bbsim, could this be done with replicas as opposed to a deployment for each?

File: voltha (hunk 4 above)
Comment (uottremb) [reply voltha]: hmm, in principle it works for more instances as 10, I think to set a hard range in the code, can be made but must not make ;)
I can define a bash constant, which is not changable, will be better ?

File: voltha (hunk 5 above)
Comment (uottremb) [reply voltha]: you are right! will update it

File: voltha (hunk 2 above)
Comment (uottremb) [reply voltha]: ok

File: voltha (hunk 4 above)
Comment (dbainbri-ciena) [reply voltha]: if it is not a hard max, then why have a max at all? What is the value is having a max?

File: voltha (hunk 6 above)
Comment (uottremb) [reply voltha]: hmm, what am I missing? if I create 5 instances of bbsim (bbsim, bbsim1, ...) during startup of the voltha script and during voltha down, all created bbsim instancses will be deleted.

Comment (davidkbainbridge) [reply voltha]: per the look below, every deployment will have an index after it, so `bbsim` will never exist it will also be `bbsim1` ... which means if you create 10 you gret `bbsim1` to `bbsim10`, correct?

File: voltha (hunk 4 above)
Comment (uottremb) [reply voltha]: a hard number in the code is always so final and I dont like it ;) 
but nevertherless, I can use a bash constant in the script.

File: voltha (hunk 6 above)
Comment (uottremb) [reply voltha]: no, first name of bbsim is untouched, is 'bbsim', the following are with numbers, like bbsim1, ..., bbsim9.

File: voltha (hunk 4 above)
Comment (dbainbri-ciena) [reply voltha]: why have a max at all, if there is really no restriction? And if they override the MAX then the code below doesn't work. EXPECT BBSIM1 ...

Why Not eliminate the MAX and for export use BBSIM*

File: voltha (hunk 6 above)
Comment (dbainbri-ciena) [reply voltha]: ok, i see that, but this seems a bit more complicated. i would use bbsim1 ... bbsimN

Comment (uottremb) [reply voltha]: hmm, I dont want change the first name, maybe existing testcases has to be then too ? This I would like prevent

Comment (uottremb) [reply voltha]: I mean, "has to be changed too"

File: voltha (hunk 1 above)
Comment (uottremb) [reply voltha]: will change this

File: voltha (hunk 4 above)
Comment (uottremb) [reply voltha]: I dont know if there exist a restriction of BBSIM instances, a meaningful using value was defined to 10 instances. I will change the setting of the value "MAX_NUM_OF_BBSIM", the value will not overwrittable anymore.

**Comment by dbainbri-ciena** [comment]
I just tested this patchset on ubuntu and it simply didn't work. I set `NUM_OF_BBSIM=3` and I still only got a single instance.

File: voltha (hunk 6 above)
Comment (uottremb) [reply voltha]: hmm, I am not with you. We have defined that the max bbsim instances are equal to 10, in the loop will searched for possible helm instances and only if the instance exist a helm remove will be called. this is the task here.

File: voltha [hunk 7]
```diff
@@ -1235,13 +1249,22 @@ if [ $WITH_BBSIM == "yes" ]; then
...
-        espin - $VERIFIED
-    fi
-    wait_for_pods - "voltha" 1 "includes" -1 "Waiting for BBSIM to start" "bbsim-.*"
+	for instance in $(seq 0 $(($NUM_OF_BBSIM-1)));
+	do
+		if [ $instance -eq 0 ]; then
+			instance_num=""
+		else
+			instance_num=$instance
+		fi
```
[the remark this answers is not in the document]
Comment (uottremb) [reply voltha]: to the first comment: this is not what I have changed, this is new for me.
to the seconed one: "would eliminate the special case for instance 1."  --> is the reason for the first instance! As I already said, I am not sure if the existing test reference to this name.

File: voltha [hunk 8]
```diff
@@ -108,6 +110,18 @@ if [ $HOSTARCH == "x86_64" ]; then
...
+# note: instances will be numbered from 0 to 9
+if [ "$1" == "up" ]; then
+	if [ "$NUM_OF_BBSIM" -lt 1 ]; then 
+			echo -e "   ${RED}Setting of BBSIM instances to minimum instances of 1!${NORMAL}"
+			NUM_OF_BBSIM=1
+	fi
+	if [ "$NUM_OF_BBSIM" -gt $(($MAX_NUM_OF_BBSIM)) ]; then 
+			echo -e "   ${RED}Setting of BBSIM instances to max supported instances of $MAX_NUM_OF_BBSIM!${NORMAL}"
+			NUM_OF_BBSIM=$MAX_NUM_OF_BBSIM
+	fi
```
[the remark this answers is not in the document]
Comment (uottremb) [reply voltha]: hmm, I thought about too. Here we have different phylosophies, one stops, the other corrects and continues. 
**Are there any rules how the applications have to behave ?**

File: voltha [hunk 9]
```diff
@@ -1235,13 +1249,22 @@ if [ $WITH_BBSIM == "yes" ]; then
...
     bspin - "Verify BBSIM Installed"
-    if [ $(helm list --deployed --short --namespace voltha "^bbsim\$" | wc -l) -ne 1 ]; then
-        espin - $NOT_VERIFIED
-        helm_install - voltha bbsim $VOLTHA_BBSIM_CHART $VOLTHA_BBSIM_CHART_VERSION "Install BBSIM"
-    else
-        espin - $VERIFIED
-    fi
-    wait_for_pods - "voltha" 1 "includes" -1 "Waiting for BBSIM to start" "bbsim-.*"
+	for instance in $(seq 0 $(($NUM_OF_BBSIM-1)));
+	do
```
[the remark this answers is not in the document]
Comment (uottremb) [reply voltha]: I can do it.

File: voltha (hunk 8 above)
[the remark this answers is not in the document]
Comment (dbainbri-ciena) [reply voltha]: The precedent is already in the code, that if a parameter is invalid (i.e. out of range) an error is reported and the applications halts.

File: voltha (hunk 7 above)
[the remark this answers is not in the document]
Comment (dbainbri-ciena) [reply voltha]: Essentially the special case for the first instance has to remain else the test cases will fail.

File: voltha (hunk 6 above)
Comment (dbainbri-ciena) [reply voltha]: Yes, but if you leave it they way it  is, then if you change the max you also have to have the array off possibilities and the code is much harder to maintain. With the suggested change the code works regardless of the max num of bbsims allowed.

File: voltha [hunk 10]
```diff
@@ -108,6 +110,18 @@ if [ $HOSTARCH == "x86_64" ]; then
...
 fi
 
+# check number (range) of bbsim, max bbsim must not exceed 10 instances!
+# note: instances will be numbered from 0 to 9
+if [ "$1" == "up" ]; then
+	if [ "$NUM_OF_BBSIM" -lt 1 ]; then 
+			echo -e "   ${RED}Setting of BBSIM instances to minimum instances of 1!${NORMAL}"
+			NUM_OF_BBSIM=1
+	fi
+	if [ "$NUM_OF_BBSIM" -gt $(($MAX_NUM_OF_BBSIM)) ]; then
```
[the remark this answers is not in the document]
Comment (uottremb) [reply voltha]: take over it.

[the remark this answers is not in the document]
Comment (dbainbri-ciena) [reply voltha]: If you no longer want to submit this patch set that is fine with me, I can close this pull request. Is that what you are asking?

File: voltha (hunk 6 above)
Comment (uottremb) [reply voltha]: I can do an extra loop to put the bbsim instances into the array EXPECT

**Comment by dbainbri-ciena** [comment]
The suggested change is cleaner than putting each possible bbsim# into the `EXPECT` array.

________________________________

File: voltha (hunk 8 above)
[the remark this answers is not in the document]
Comment (uottremb) [reply voltha]: ok, you are the boss! I will change the behaviour.

File: voltha (hunk 10 above)
[the remark this answers is not in the document]
Comment (uottremb) [reply voltha]: no, I update my request.

File: voltha (hunk 6 above)
Comment (uottremb) [reply voltha]: hi again,
I am sorry, I didnt registrated that you show me a diff. I will take your suggestion.

File: voltha [hunk 11]
```diff
@@ -382,32 +396,15 @@ push_onos_config() {
...
 
-    # Thanks to the latest version of ONOS using the return code 207 this gets a
-    # whole lot nastier. Can't thank them enough for doing this. So in order to
-    # capture the command and the output in the log file as well as capture the
-    # status code to verify it is 200 and not 207 mutltiple files and a bit of
-    # hackery must be used. Thanks again ONOS.
-    local CMD_ECHO=$(mktemp -u)
-    local CMD_OUTPUT=$(mktemp -u)
-    local SC_OUTPUT=$(mktemp -u)
-
```
Comment (dbainbri-ciena) [inline voltha]: Why is this removed?

File: voltha [hunk 12]
```diff
@@ -382,32 +396,15 @@ push_onos_config() {
...
     bspin - "$MSG $GEAR"
     while true; do
         if [ $TYPE == "file" ]; then
-          (set -x; curl --fail -sSL --user karaf:karaf -w "%{http_code}" -o $CMD_OUTPUT -X POST -H Content-Type:application/json http://127.0.0.1:$ONOS_API_PORT/onos/v1/$RESOURCE --data @$DATA >$SC_OUTPUT 2>/dev/null) >>$CMD_ECHO 2>&1
```
Comment (dbainbri-ciena) [inline voltha]: why is this removed

File: voltha [hunk 13]
```diff
@@ -382,32 +396,15 @@ push_onos_config() {
...
+          (set -x; curl --fail -sSL --user karaf:karaf -X POST -H Content-Type:application/json http://127.0.0.1:$ONOS_API_PORT/onos/v1/$RESOURCE --data @$DATA >>$LOG 2>&1) >>$LOG 2>&1
         fi
         if [ $TYPE == "json" ]; then
-          (set -x; curl --fail -sSL --user karaf:karaf -w "%{http_code}" -o $CMD_OUTPUT -X POST -H Content-Type:application/json http://127.0.0.1:$ONOS_API_PORT/onos/v1/$RESOURCE --data $DATA >$SC_OUTPUT 2>/dev/null) >>$CMD_ECHO 2>&1
```
Comment (dbainbri-ciena) [inline voltha]: why is this removed

File: voltha [hunk 14]
```diff
@@ -382,32 +396,15 @@ push_onos_config() {
...
+          (set -x; curl --fail -sSL --user karaf:karaf -X POST -H Content-Type:application/json http://127.0.0.1:$ONOS_API_PORT/onos/v1/$RESOURCE --data $DATA >>$LOG 2>&1) >>$LOG 2>&1
         fi
-        RESULT=$?
-        # Dump everything to the log
-        cat $CMD_ECHO >> $LOG
-        cat $CMD_OUTPUT >> $LOG
-        SC=$(cat $SC_OUTPUT)
-
-        # clean up temp files
```
Comment (dbainbri-ciena) [inline voltha]: why is this removed

Comment (uottremb) [reply voltha]: I dont know, maybe there was change within the 4 days? This is my first time where I work with github and I think something goes wrong during our discussion ;)
I will start a new pull request, ok ?

**Comment by uottremb** [comment]
I will start new with a fresh repo.
